### DISCUSSION TITLE
Frequently Used Transactions

### DISCUSSION THREAD

**Discussion by DevAnyKR** [moneymanagerex/moneymanagerex#5867] [Ideas]
Hopes...
I have a few Frequently Used Transactions(FUT).
Once I have registered it, simply select FUT Item from the New Transaction form and the content will be automatically fill.

Do you plan to add this feature or is it possible to add it?
(This feature does not copy existing transactions. Because it takes time to search for the original.)
And the contents of the notes may be the same, and even if they are different, the forms are similar.

Detail...
The Favorite Transactions List appears in the list by Type selection.
Amount, Account, Payee (From or To), Category, Notes, others... items automatically populate

sample image...
![2023-04-28 111842](https://user-images.githubusercontent.com/110871727/235040068-bd98a142-aa81-4824-8b7d-580260c9b75c.png)

Organize Frequently Used Transactions
![2023-04-28 113805](https://user-images.githubusercontent.com/110871727/235041154-2be5307a-a143-4e4e-bd91-a870b7ee632a.png)

**Comment by tactilis** [comment]
@DevelopmentAnything A similar feature has been discussed previously. Please see https://github.com/moneymanagerex/moneymanagerex/issues/4901

That issue is still open but there hasn't been any enthusiasm to develop it along the lines of what you are requesting.

The concluding remark was https://github.com/moneymanagerex/moneymanagerex/issues/4901#issuecomment-1206682756:

> To be fair with your suggestion you are gaining no more than 1 click for the payee, because if you use the "last category used for payee", this is autofilled as well and the account depends where you right click and add new transaction. IMHO creating another window with other logic etc.. for such small advantage it's not really useful.

<br>

A simple solution requiring no development effort that you could use right now was made here: https://github.com/moneymanagerex/moneymanagerex/issues/4901#issuecomment-1204126353

In summary, that approach is:

1. Select the Temp account containing your template Frequently Used Transactions.
2. Ctrl+C to copy the transaction you want.
3. Select the account where you want the new transaction to be created.
4. Ctrl+V to paste the template transaction.

**Reply by DevAnyKR** [reply]
The problem with this is that it's exposed to the entire transaction. Despite the closed setting.

**Comment by n-stein** [comment]
I suppose by definition a "Frequently Used Transaction" is one that is recurring (though perhaps not on a set schedule), so another option is to use the native "Recurring Transactions" functionality as mentioned here [#4901 (comment)](https://github.com/moneymanagerex/moneymanagerex/issues/4901#issuecomment-1206680783):

1. Create a manual recurring transaction for each of your frequently used transactions.
2. Any time you want to enter it, go to recurring transactions screen, right click on the transaction and select "Enter next Occurrence..."

Note this would prompt you with the transaction dialog pre-filled where you can make changes to amount, notes, etc whereas copying/pasting will duplicate the transaction as-is in the target account. Depending on your use case this could be beneficial.

**Reply by DevAnyKR** [reply]
If  enter a recurring transaction, the recurring transaction item disappears.

**Reply by DevAnyKR** [reply]
1. New Recurring Transaction
![스크린샷 2023-04-30 152914](https://user-images.githubusercontent.com/110871727/235339237-f6a8c460-6531-46b9-985b-9f0edf9967d2.png)

2. List
![스크린샷 2023-04-30 152928](https://user-images.githubusercontent.com/110871727/235339238-0c672dd3-1c84-4304-a429-df1504b7d0d8.png)

3. Enter Recurring Transaction
![스크린샷 2023-04-30 152947](https://user-images.githubusercontent.com/110871727/235339239-86db740a-94f6-4e59-820c-5a3f989ce753.png)

4. Disappeared from the list
![스크린샷 2023-04-30 152955](https://user-images.githubusercontent.com/110871727/235339240-b9f63757-8fed-4ce8-864e-ce2b7ebea62c.png)

This feature is not available for 'Frequently Used Transactions'.

**Comment by vomikan** [comment]
Just copy then past already present transection in any account then edit details.

**Reply by DevAnyKR** [reply]
I've already mentioned it.
This is because it takes time to retrieve the source.

I can use that function with a temporary account but selecting my favorite transactions on the transaction input screen of a account is a feature used by many account book app in Korea. It is possible with the above two answers.

**Comment by DevAnyKR** [comment]
The above method is difficult to use continuously.
It seems necessary to add a separate 'Frequently Used Transaction' function.

1. This transaction is not a recurring transaction in that cycle.
2. Use it anytime. Past, present, future all.
3. can only register multiple times a day or once a month.
4. must complete the task at once on the transaction registration screen. Not copy and paste.
5. The amount and content may be the same or different.

**Comment by vomikan** [comment]
I do not recommend using recurring transactions, because this will violate the Cash flow report.

**Reply by n-stein** [reply]
Setting the transaction status to "Void" would prevent any impact to cash flow reporting.

**Reply by DevAnyKR** [reply]
I hope the record is not deleted by the Enter (Next) button in the recurring transaction. How do I set it up so it won't be deleted?
repeats = none condition.

**Reply by n-stein** [reply]
Since it is manual you can pick any "Repeats" condition you want, the only thing it will determine is what transaction date is initially populated. You could pick "Daily", you would just need to adjust the "Date Paid" when you enter the transaction.

**Reply by nmsleal** [reply]
The only problem with the "daily" approach is that there'll be lots of recurring tasks to advance if, for any reason, we take a few days between two Frequently Used Transactions... I think that a kind of a "permanent flag" (to prevent deletion), combined with a "void" type and a new kind of repeat "Reusable Adhoc", for example, would fit this purpose, allowing us to use the Recurring Transactions mechanism and, simultaneously, not being such an huge development effort.
This feature is very interesting. I frequently find myself advancing lots of poorly recurrent transactions setup as twice a week or other recurrence.

**Comment by DevAnyKR** [comment]
This is a frequently used transaction registration function of my app for MMEX.

1. Organize FavoriteUsedTransaction(=Transaction Preset)
![스크린샷 2023-05-02 180646](https://user-images.githubusercontent.com/110871727/235677798-c3c3de8c-3c18-498c-8e8a-f78fd86cb531.png)

2. New Transaction Dialog and Select Preset Item
![스크린샷 2023-05-02 221301](https://user-images.githubusercontent.com/110871727/235679475-014a0069-4c81-4647-a23d-7f403dd393ed.png)
All on one screen.

3. Success
![스크린샷 2023-05-02 222158](https://user-images.githubusercontent.com/110871727/235678861-70437fcd-733c-428b-a15f-c313de7c7f84.png)

**Comment by DevAnyKR** [comment]
If I use scheduled transaction as https://github.com/moneymanagerex/moneymanagerex/discussions/6568#discussioncomment-8716126 , manually entering it will delete the scheduled transaction, which means it cannot be used as a frequently used transaction.
We need to improve both scheduled transactions and frequently favorite transactions so that they can be easily used.

**Reply by DevAnyKR** [reply]
Frequently Used Transactions is ...
"Transaction Template "

ex) Aliexpress scheduled transaction

```
account: xx Card
amount: 0
payee: aliexpress
category: product

Note:

blahblah

Qty: 1 ea
sum: $0
delivery: $0
Coin: -$0
discount: -$0
Total: $0
```

-----

![image](https://github.com/moneymanagerex/moneymanagerex/assets/110871727/5d84d2c9-3aa3-48ee-b2e8-59ad00076fce)

and manually use it...

![image](https://github.com/moneymanagerex/moneymanagerex/assets/110871727/4d33ead8-9887-413c-abb4-434c1dc9ca3b)

enter scheduled transaction dialog, click OK

deleted scheduled transaction 😱😱😱 can not use for Transaction Template(Frequently Used Transactions)

**Comment by georgeef** [comment]
a simple way to implement what you described:
- use "All Transactions" and select "View All Transactions" (i.e., no time limit)
- add a few "frequently used transactions" with date in the far future (e.g., 2100-01-01)
- all these "frequently used transactions" appear at the bottom and they are shadowed, so easy to distinguish
- all these "frequently used transactions" do not affect cash flow (they do, but after 2100) 
- if you have only a few "frequently used transactions" it is easy to find one of them; look at the bottom
- when you want to enter one of them, select it and press Duplicate
- keep the default setting for the duplicate transaction date (do not use original date)
- press OK and the transactions is added with the date of today; the template stays at the bottom for next time
- if you don't want to see the template transactions, enable "Ignore Future Transactions" from menu
- if shadowing of templates is not enough, add color, but you have to clear the color (with bindkey) after duplicate

there is one problem though: the template transactions affect the balance in Dashboard, which is of course annoying. you can make them Void, but then you need to set the status on every duplicate, which is also annoying. "Ignore Future Transactions" has no effect on Dashboard; anyway this wouldn't be convenient, even if it did, because it affects the panels at the same time. i cannot think of a good solution, but an easy fix would be to have an option to ignore future transactions in Dashboard, independently of the corresponding option for panels. anyway it would be good to have such an option, even without "frequently used transactions" (or better "far future transactions").

**Reply by vomikan** [reply]
I like this idea.
You can make a special status for transactions 'T' for example. 
Such a transaction can be undated (which means as now). It cannot be deleted except in a special way. It does not affect statistics. double-clicking on it leads to the creation of a new transaction.

**Reply by georgeef** [reply]
The date is not so important, because Duplicate anyway resets the date. An empty date may create problems in several places in the code. Better to have some dummy date than empty. It could be a special date (like 9999-12-12), but not empty. Still code should be able to properly handle dates with special meaning.

A status which does not affect balance is a good idea, but we have Void already for this purpose, and as described above this is inconvenient for this use case. It could be another flag. All this requires a DB change.

**Reply by georgeef** [reply]
another idea (requires changes,  but not DB changes):
- dates starting with `9999` are special.
- transactions with special date do not count in any account balance.
- scheduled transactions with special date do not execute automatically (anyway they are scheduled for year 9999, so no need to change something).
- add a "pin" option for transactions with special date: they are either visible in applicable panels with any filter (if the panel is sorted by date, they appear at the bottom), or they are not visible with any filter. this options can be named "Show Template Transactions", with default False, and persistent (stored) per panel.
- scheduled transactions with special date are pinned the same way. if they are visible and their panel is sorted by date, they appear at the bottom, or at the top if sorting is inverted (so it is easy to find them, and they have fixed order).
- when a scheduled transaction with special date is entered, its dates are not incremented. effectively 9999 means infinity, and infinity cannot be increased further.

support of special dates has the following benefits:
- the functionality of filters in panels is independent of the visibility of special dates.
- "Ignore Future Transactions" is independent of special dates (with any setting, special dates are always ignored in any balance).
- transactions are classified as: past (they ALWAYS count in balance), future (they MAY count in balance), and far future (9999, they NEVER count in balance).
- other code except the special handling listed above, works fine with special dates, because they have the format of a normal date, they are comparable, etc.

the users have two options to implement templates (like frequently used transactions):
- add transactions with date starting with 9999. their visibility can be toggled with a bindkey in transactions panels. press Duplicate to copy them.
- add scheduled transactions with date starting with 9999.  they are always visible there. press Enter to copy them. the countdown is as for all other transactions, but they don't change date after execution.

another extension would be that dates starting with 0000 are also special, they mean -infinity, and they are handled like above. they appear at the top of the list (instead of bottom for 9999) when sorted by date. i think the bring little additional benefit for the effort to implement their special handling.

**Reply by georgeef** [reply]
@vomikan, @whalley 
What do you think about the idea above to use year 9999 to implement template transactions? This year is not used in practice, or when used it mostly has the semantics of "temporary" or "template". No DB change is needed, and the semantics of DB are largely unaffected.

Impact:
Most of the code will continue to treat year 9999 as normal date. The calculation of balances is affected, but in a way that makes sense (year 9999 is most probably a mistake if it appears in normal transactions). The new semantic interpretation is: future is the period between today and before year 9999; year 9999 is in future but unreachable (infinite), so it doesn't count in any balance. The month and day in year 9999 give some flexibility to the user to sort template transactions.

Changes:
There is some effort (relatively small) to support the concept of templates in the panels and to add options to the user.

EDIT 1: Reports are also affected: they should ignore transactions or scheduled transactions with year 9999.

EDIT 2: It is assumed that the maximum year is 9999. If the user enters a year greater or equal to 1000, it is stored as 9999. This simplifies the check for special dates: if `DATE` is a date string, then the conditions `DATE >= 9999` indicates if `DATE` is special.

**Reply by vomikan** [reply]
I did not understand where such templates will be stored? In the transaction table or in scheduled transactions?

**Reply by georgeef** [reply]
In either. It is up to the user to choose. Users may want to create template transactions, in order to see them in the transactions panel and use them from there. Or they may create template scheduled transactions, because they don't want to have template entries in the transactions table.
The main rule is: transactions or scheduled transactions with year 9999 do not count in any balance and they are ignored in reports.

**Reply by whalley** [reply]
Not keen on messing with the date. Why not use another status value, e.g. "T" for Template, to distinguish this.

**Reply by georgeef** [reply]
The status is part of the template, if we set it to "T" then users need to configure it every time they use (duplicate) the template. On the other hand, the date is not needed in the template; when a template is used (duplicated), the date is reset to today.

I agree that a new status is more clean solution from model point of view, although not very convenient for the user.

What about having more new status options like:
- T: template for Unreconciled
- TR: template for Reconciled
- TV, TF, TD: similar

When a template is used, the status will be set to the corresponding status without T. This would double the number of status options. The interface dialog could make it a little simpler, by providing a checkbox for Template, so from the user point of view it appears as if T is a new flag independent of Status.

Change in the status options is essentially a change in DB, correct?  At least the Android app must be compatible.

**Reply by whalley** [reply]
> Change in the status options is essentially a change in DB, correct? 

At least the Android app must be compatible.
Strictly spekaing yes it is, even if the structure has not changed. It may be better therefore to add a 'Template' flag to the DB rather than overload existing fields like date or status.

**Comment by georgeef** [comment]
if you use scheduled transactions to emulate "frequently used transactions", you will have a few problems:
- you need to switch back and forth between transactions panel and scheduled transactions panel (there is a PR which shows the scheduled transactions in the transactions panel - not merged yet).
- the scheduled transactions have a repetition period, which cannot be zero. after each usage (Enter) the scheduled transaction moves to the next occurrence, which is a little strange for your use case.
- if you have many scheduled transactions, it is not easy to find the one that you need (although it is easier than in the panel). you may want to color them and un-color after Enter.
- the scheduled transactions list rotates all the time, so you cannot memorize the positions of the frequently used transactions (compared to having them all at the bottom of a panel, in fixed order).
